{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from copy import deepcopy\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_path = 'GCA_000046845.1_ASM4684v1_cds_from_genomic.fna'\n",
    "genbank_path = 'GCA_000046845.1_ASM4684v1_genomic.gbff'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>tag</th>\n",
       "      <th>target_id</th>\n",
       "      <th>LB0R1</th>\n",
       "      <th>LB0R2</th>\n",
       "      <th>LB0R3</th>\n",
       "      <th>LB1R1</th>\n",
       "      <th>LB1R2</th>\n",
       "      <th>LB1R3</th>\n",
       "      <th>M9LQR1</th>\n",
       "      <th>M9LQR2</th>\n",
       "      <th>M9LQR3</th>\n",
       "      <th>M9SFR1</th>\n",
       "      <th>M9SFR2</th>\n",
       "      <th>M9SFR3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>dnaA</td>\n",
       "      <td>ACIAD0001</td>\n",
       "      <td>lcl|CR543861.1_cds_CAG66986.1_1</td>\n",
       "      <td>298.2740</td>\n",
       "      <td>222.8160</td>\n",
       "      <td>306.2610</td>\n",
       "      <td>302.868</td>\n",
       "      <td>300.4660</td>\n",
       "      <td>307.0410</td>\n",
       "      <td>60.4627</td>\n",
       "      <td>59.4571</td>\n",
       "      <td>48.8814</td>\n",
       "      <td>154.0610</td>\n",
       "      <td>135.4880</td>\n",
       "      <td>164.6930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dnaN</td>\n",
       "      <td>ACIAD0002</td>\n",
       "      <td>lcl|CR543861.1_cds_CAG66987.1_2</td>\n",
       "      <td>302.3590</td>\n",
       "      <td>253.9690</td>\n",
       "      <td>292.0170</td>\n",
       "      <td>366.812</td>\n",
       "      <td>265.3100</td>\n",
       "      <td>316.2670</td>\n",
       "      <td>29.5314</td>\n",
       "      <td>55.0744</td>\n",
       "      <td>57.5752</td>\n",
       "      <td>189.6850</td>\n",
       "      <td>151.6120</td>\n",
       "      <td>227.9460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>recF</td>\n",
       "      <td>ACIAD0003</td>\n",
       "      <td>lcl|CR543861.1_cds_CAG66988.1_3</td>\n",
       "      <td>254.4090</td>\n",
       "      <td>235.6500</td>\n",
       "      <td>180.6980</td>\n",
       "      <td>155.834</td>\n",
       "      <td>226.5090</td>\n",
       "      <td>203.4650</td>\n",
       "      <td>87.8677</td>\n",
       "      <td>69.7656</td>\n",
       "      <td>75.6724</td>\n",
       "      <td>50.6984</td>\n",
       "      <td>55.3793</td>\n",
       "      <td>59.9481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gyrB</td>\n",
       "      <td>ACIAD0004</td>\n",
       "      <td>lcl|CR543861.1_cds_CAG66989.1_4</td>\n",
       "      <td>582.1220</td>\n",
       "      <td>543.3160</td>\n",
       "      <td>423.3880</td>\n",
       "      <td>438.286</td>\n",
       "      <td>618.6660</td>\n",
       "      <td>383.9690</td>\n",
       "      <td>90.8570</td>\n",
       "      <td>98.7740</td>\n",
       "      <td>114.1330</td>\n",
       "      <td>149.3250</td>\n",
       "      <td>150.0720</td>\n",
       "      <td>158.6460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ACIAD0005</td>\n",
       "      <td>lcl|CR543861.1_cds_CAG66990.1_5</td>\n",
       "      <td>48.0218</td>\n",
       "      <td>59.8179</td>\n",
       "      <td>57.9693</td>\n",
       "      <td>146.440</td>\n",
       "      <td>60.4595</td>\n",
       "      <td>63.1651</td>\n",
       "      <td>1780.5700</td>\n",
       "      <td>2163.6600</td>\n",
       "      <td>1451.0800</td>\n",
       "      <td>269.3740</td>\n",
       "      <td>476.8390</td>\n",
       "      <td>277.2980</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene        tag                        target_id     LB0R1     LB0R2  \\\n",
       "0  dnaA  ACIAD0001  lcl|CR543861.1_cds_CAG66986.1_1  298.2740  222.8160   \n",
       "1  dnaN  ACIAD0002  lcl|CR543861.1_cds_CAG66987.1_2  302.3590  253.9690   \n",
       "2  recF  ACIAD0003  lcl|CR543861.1_cds_CAG66988.1_3  254.4090  235.6500   \n",
       "3  gyrB  ACIAD0004  lcl|CR543861.1_cds_CAG66989.1_4  582.1220  543.3160   \n",
       "4   NaN  ACIAD0005  lcl|CR543861.1_cds_CAG66990.1_5   48.0218   59.8179   \n",
       "\n",
       "      LB0R3    LB1R1     LB1R2     LB1R3     M9LQR1     M9LQR2     M9LQR3  \\\n",
       "0  306.2610  302.868  300.4660  307.0410    60.4627    59.4571    48.8814   \n",
       "1  292.0170  366.812  265.3100  316.2670    29.5314    55.0744    57.5752   \n",
       "2  180.6980  155.834  226.5090  203.4650    87.8677    69.7656    75.6724   \n",
       "3  423.3880  438.286  618.6660  383.9690    90.8570    98.7740   114.1330   \n",
       "4   57.9693  146.440   60.4595   63.1651  1780.5700  2163.6600  1451.0800   \n",
       "\n",
       "     M9SFR1    M9SFR2    M9SFR3  \n",
       "0  154.0610  135.4880  164.6930  \n",
       "1  189.6850  151.6120  227.9460  \n",
       "2   50.6984   55.3793   59.9481  \n",
       "3  149.3250  150.0720  158.6460  \n",
       "4  269.3740  476.8390  277.2980  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('TPM_matrix.csv')\n",
    "locus_tags = list(df.tag)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get_interregions\n",
    "\n",
    "seq_record = SeqIO.read(open(genbank_path,'r'), 'genbank')\n",
    "genome_size = len(seq_record)\n",
    "cds_list = []\n",
    "# Loop over the gnome, get CDS features on each of the strands (locations)\n",
    "for feature in seq_record.features:\n",
    "    if feature.type == 'CDS':\n",
    "        mystart = feature.location.start.position\n",
    "        myend = feature.location.end.position\n",
    "        strand = feature.strand\n",
    "        locus_tag = feature.qualifiers['locus_tag'][0] # each cds should have only a single tag\n",
    "        cds_list.append([mystart,myend,strand,locus_tag])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fwd_start_and_end(this_start,curr_cds,strand):\n",
    "    init_cds = deepcopy(curr_cds)\n",
    "    for ii in range(curr_cds,-1,-1): # get nearest upstream cds on same strand\n",
    "        if cds_list[ii][2] == strand:\n",
    "            last_end = cds_list[ii][1]\n",
    "            if curr_cds == (len(cds_list)-1) and this_start < last_end: \n",
    "                if genome_size - last_end + this_start > 40: \n",
    "                    return this_start, last_end\n",
    "            elif curr_cds != (len(cds_list)-1) and this_start - last_end > 40:\n",
    "                return this_start, last_end\n",
    "            else: \n",
    "                this_start = cds_list[ii][0]\n",
    "                curr_cds = ii - 1\n",
    "                this_start, last_end = get_fwd_start_and_end(this_start,curr_cds,strand)\n",
    "                \n",
    "def get_rev_start_and_end(this_start,curr_cds,strand):\n",
    "    for ii in range(curr_cds,len(cds_list)): # get nearest downstream cds on same strand\n",
    "        if cds_list[ii][2] == strand:\n",
    "            last_end = cds_list[ii][0]\n",
    "            if last_end - this_start > 40: \n",
    "                return this_start, last_end\n",
    "            else: \n",
    "                this_start = cds_list[ii][1]\n",
    "                curr_cds = ii + 1\n",
    "                if curr_cds == len(cds_list) or ii == len(cds_list): # then we have exhausted the downstream coding sequences on this strand\n",
    "                    return this_start, last_end\n",
    "                else: \n",
    "                    this_start, last_end = get_rev_start_and_end(this_start,curr_cds,strand)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get putative promoter region for each CDS\n",
    "intergenic_records = []\n",
    "\n",
    "# forward strand\n",
    "for ii in range(len(cds_list)):\n",
    "    strand = cds_list[ii][2]\n",
    "    if strand == 1: # only look upstream for intergenic region\n",
    "        \n",
    "        if ii == 0: \n",
    "            this_start = cds_list[ii][0]\n",
    "            this_start, last_end = get_fwd_start_and_end(this_start,len(cds_list)-1,strand)\n",
    "            if this_start < last_end: \n",
    "                intergene_seq = seq_record.seq[last_end:genome_size] + seq_record.seq[0:this_start]\n",
    "            else: \n",
    "                intergene_seq = seq_record.seq[last_end:this_start]\n",
    "            if len(intergene_seq) > 1000: \n",
    "                intergene_seq = intergene_seq[-1000:]\n",
    "            intergenic_records.append(SeqRecord(intergene_seq,id=cds_list[ii][3],\n",
    "                         description='%d,%d,%d'%(last_end,this_start,strand)))  \n",
    "        \n",
    "        if ii > 0:\n",
    "            this_start = cds_list[ii][0]\n",
    "            this_start, last_end = get_fwd_start_and_end(this_start,ii-1,strand) # recursively search for intergenic region on same strand\n",
    "            intergene_seq = seq_record.seq[last_end:this_start]\n",
    "            if len(intergene_seq) > 1000: \n",
    "                intergene_seq = intergene_seq[-1000:]\n",
    "            intergenic_records.append(SeqRecord(intergene_seq,id=cds_list[ii][3],\n",
    "                         description='%d,%d,%d'%(last_end,this_start,strand)))                \n",
    "                    \n",
    "\n",
    "# reverse strand\n",
    "for ii, features in reversed(list(enumerate(cds_list))):               \n",
    "    strand = cds_list[ii][2]\n",
    "    if strand == -1: # only look downstream for intergenic region\n",
    "        if ii < len(cds_list)-1:\n",
    "            this_start = cds_list[ii][1]\n",
    "            this_start, last_end = get_rev_start_and_end(this_start,ii+1,strand)\n",
    "            \n",
    "            if last_end < this_start: \n",
    "                intergene_seq = seq_record.seq[this_start:genome_size] + seq_record.seq[0:last_end]\n",
    "            else:\n",
    "                intergene_seq = seq_record.seq[this_start:last_end]\n",
    "            if len(intergene_seq) > 1000:\n",
    "                intergene_seq = intergene_seq[0:1000]\n",
    "            intergenic_records.append(SeqRecord(intergene_seq.reverse_complement(),id=cds_list[ii][3],\n",
    "                         description='%d,%d,%d'%(last_end,this_start,strand)))                \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# place records in dataframe\n",
    "\n",
    "intergenic_record_dict = {'locus_tag':[],'start':[],'end':[],'strand':[],'seq':[]}\n",
    "for record in intergenic_records: \n",
    "    intergenic_record_dict['locus_tag'].append(record.id)\n",
    "    mystart, myend, strand = record.description.split(',')\n",
    "    mystart, myend, strand = int(mystart), int(myend), int(strand)\n",
    "    intergenic_record_dict['start'].append(mystart)\n",
    "    intergenic_record_dict['end'].append(myend)    \n",
    "    intergenic_record_dict['strand'].append(strand)    \n",
    "    intergenic_record_dict['seq'].append(str(record.seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "intergenic_record_df = pd.DataFrame(intergenic_record_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# intergenic_record_df.to_csv('intergenic_records.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
