{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from copy import deepcopy\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_path = 'GCF_000046845.1_ASM4684v1_cds_from_genomic.fna'\n",
    "genbank_path = 'GCF_000046845.1_ASM4684v1_genomic.gbff'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>tag</th>\n",
       "      <th>target_id</th>\n",
       "      <th>LB0R1</th>\n",
       "      <th>LB0R2</th>\n",
       "      <th>LB0R3</th>\n",
       "      <th>LB1R1</th>\n",
       "      <th>LB1R2</th>\n",
       "      <th>LB1R3</th>\n",
       "      <th>M9LQR1</th>\n",
       "      <th>M9LQR2</th>\n",
       "      <th>M9LQR3</th>\n",
       "      <th>M9SFR1</th>\n",
       "      <th>M9SFR2</th>\n",
       "      <th>M9SFR3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>dnaA</td>\n",
       "      <td>ACIAD_RS00005</td>\n",
       "      <td>lcl|NC_005966.1_cds_WP_004930068.1_1</td>\n",
       "      <td>254.4710</td>\n",
       "      <td>195.2650</td>\n",
       "      <td>258.4730</td>\n",
       "      <td>275.958</td>\n",
       "      <td>240.0250</td>\n",
       "      <td>217.3340</td>\n",
       "      <td>56.1754</td>\n",
       "      <td>55.1445</td>\n",
       "      <td>45.2804</td>\n",
       "      <td>147.637</td>\n",
       "      <td>127.7610</td>\n",
       "      <td>157.6420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dnaN</td>\n",
       "      <td>ACIAD_RS00010</td>\n",
       "      <td>lcl|NC_005966.1_cds_WP_004930066.1_2</td>\n",
       "      <td>257.9470</td>\n",
       "      <td>222.5650</td>\n",
       "      <td>246.4550</td>\n",
       "      <td>334.218</td>\n",
       "      <td>211.9420</td>\n",
       "      <td>223.8630</td>\n",
       "      <td>27.4375</td>\n",
       "      <td>51.0805</td>\n",
       "      <td>53.3345</td>\n",
       "      <td>181.774</td>\n",
       "      <td>142.9680</td>\n",
       "      <td>218.1870</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>recF</td>\n",
       "      <td>ACIAD_RS00015</td>\n",
       "      <td>lcl|NC_005966.1_cds_WP_004930061.1_3</td>\n",
       "      <td>217.0360</td>\n",
       "      <td>206.5120</td>\n",
       "      <td>152.5050</td>\n",
       "      <td>141.987</td>\n",
       "      <td>180.9460</td>\n",
       "      <td>144.0180</td>\n",
       "      <td>81.6377</td>\n",
       "      <td>64.7067</td>\n",
       "      <td>70.0990</td>\n",
       "      <td>48.584</td>\n",
       "      <td>52.2221</td>\n",
       "      <td>57.3814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gyrB</td>\n",
       "      <td>ACIAD_RS00020</td>\n",
       "      <td>lcl|NC_005966.1_cds_WP_004930060.1_4</td>\n",
       "      <td>496.6680</td>\n",
       "      <td>476.1380</td>\n",
       "      <td>357.3170</td>\n",
       "      <td>399.349</td>\n",
       "      <td>494.2160</td>\n",
       "      <td>271.7890</td>\n",
       "      <td>84.4136</td>\n",
       "      <td>91.6070</td>\n",
       "      <td>105.7230</td>\n",
       "      <td>143.101</td>\n",
       "      <td>141.5100</td>\n",
       "      <td>151.8540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>ACIAD_RS00025</td>\n",
       "      <td>lcl|NC_005966.1_cds_WP_004930057.1_5</td>\n",
       "      <td>47.7092</td>\n",
       "      <td>56.7863</td>\n",
       "      <td>48.9246</td>\n",
       "      <td>133.345</td>\n",
       "      <td>55.1851</td>\n",
       "      <td>44.6848</td>\n",
       "      <td>1739.3300</td>\n",
       "      <td>2147.7500</td>\n",
       "      <td>1420.1400</td>\n",
       "      <td>292.355</td>\n",
       "      <td>458.4010</td>\n",
       "      <td>280.9240</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene            tag                             target_id     LB0R1  \\\n",
       "0  dnaA  ACIAD_RS00005  lcl|NC_005966.1_cds_WP_004930068.1_1  254.4710   \n",
       "1  dnaN  ACIAD_RS00010  lcl|NC_005966.1_cds_WP_004930066.1_2  257.9470   \n",
       "2  recF  ACIAD_RS00015  lcl|NC_005966.1_cds_WP_004930061.1_3  217.0360   \n",
       "3  gyrB  ACIAD_RS00020  lcl|NC_005966.1_cds_WP_004930060.1_4  496.6680   \n",
       "4   NaN  ACIAD_RS00025  lcl|NC_005966.1_cds_WP_004930057.1_5   47.7092   \n",
       "\n",
       "      LB0R2     LB0R3    LB1R1     LB1R2     LB1R3     M9LQR1     M9LQR2  \\\n",
       "0  195.2650  258.4730  275.958  240.0250  217.3340    56.1754    55.1445   \n",
       "1  222.5650  246.4550  334.218  211.9420  223.8630    27.4375    51.0805   \n",
       "2  206.5120  152.5050  141.987  180.9460  144.0180    81.6377    64.7067   \n",
       "3  476.1380  357.3170  399.349  494.2160  271.7890    84.4136    91.6070   \n",
       "4   56.7863   48.9246  133.345   55.1851   44.6848  1739.3300  2147.7500   \n",
       "\n",
       "      M9LQR3   M9SFR1    M9SFR2    M9SFR3  \n",
       "0    45.2804  147.637  127.7610  157.6420  \n",
       "1    53.3345  181.774  142.9680  218.1870  \n",
       "2    70.0990   48.584   52.2221   57.3814  \n",
       "3   105.7230  143.101  141.5100  151.8540  \n",
       "4  1420.1400  292.355  458.4010  280.9240  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('TPM_matrix.csv')\n",
    "locus_tags = list(df.tag)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get_interregions\n",
    "\n",
    "seq_record = SeqIO.read(open(genbank_path,'r'), 'genbank')\n",
    "genome_size = len(seq_record)\n",
    "cds_list = []\n",
    "# Loop over the gnome, get CDS features on each of the strands (locations)\n",
    "for feature in seq_record.features:\n",
    "    if feature.type == 'CDS':\n",
    "        mystart = feature.location.start.position\n",
    "        myend = feature.location.end.position\n",
    "        strand = feature.strand\n",
    "        locus_tag = feature.qualifiers['locus_tag'][0] # each cds should have only a single tag\n",
    "        cds_list.append([mystart,myend,strand,locus_tag])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fwd_start_and_end(this_start,curr_cds,strand):\n",
    "    init_cds = deepcopy(curr_cds)\n",
    "    for ii in range(curr_cds,-1,-1): # get nearest upstream cds on same strand\n",
    "        if cds_list[ii][2] == strand:\n",
    "            last_end = cds_list[ii][1]\n",
    "            if curr_cds == (len(cds_list)-1) and this_start < last_end: \n",
    "                if genome_size - last_end + this_start > 40: \n",
    "                    return this_start, last_end\n",
    "            elif curr_cds != (len(cds_list)-1) and this_start - last_end > 40:\n",
    "                return this_start, last_end\n",
    "            else: \n",
    "                this_start = cds_list[ii][0]\n",
    "                curr_cds = ii - 1\n",
    "                this_start, last_end = get_fwd_start_and_end(this_start,curr_cds,strand)\n",
    "                \n",
    "def get_rev_start_and_end(this_start,curr_cds,strand):\n",
    "    for ii in range(curr_cds,len(cds_list)): # get nearest downstream cds on same strand\n",
    "        if cds_list[ii][2] == strand:\n",
    "            last_end = cds_list[ii][0]\n",
    "            if last_end - this_start > 40: \n",
    "                return this_start, last_end\n",
    "            else: \n",
    "                this_start = cds_list[ii][1]\n",
    "                curr_cds = ii + 1\n",
    "                if curr_cds == len(cds_list) or ii == len(cds_list): # then we have exhausted the downstream coding sequences on this strand\n",
    "                    return this_start, last_end\n",
    "                else: \n",
    "                    this_start, last_end = get_rev_start_and_end(this_start,curr_cds,strand)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "intergenic_records = []\n",
    "\n",
    "# forward strand\n",
    "for ii in range(len(cds_list)):\n",
    "    strand = cds_list[ii][2]\n",
    "    if strand == 1: # only look upstream for intergenic region\n",
    "        \n",
    "        if ii == 0: \n",
    "            this_start = cds_list[ii][0]\n",
    "            this_start, last_end = get_fwd_start_and_end(this_start,len(cds_list)-1,strand)\n",
    "            if this_start < last_end: \n",
    "                intergene_seq = seq_record.seq[last_end:genome_size] + seq_record.seq[0:this_start]\n",
    "            else: \n",
    "                intergene_seq = seq_record.seq[last_end:this_start]\n",
    "            if len(intergene_seq) > 1000: \n",
    "                intergene_seq = intergene_seq[-1000:]\n",
    "            intergenic_records.append(SeqRecord(intergene_seq,id=cds_list[ii][3],\n",
    "                         description='%d,%d,%d'%(last_end,this_start,strand)))  \n",
    "        \n",
    "        if ii > 0:\n",
    "            this_start = cds_list[ii][0]\n",
    "            this_start, last_end = get_fwd_start_and_end(this_start,ii-1,strand) # recursively search for intergenic region on same strand\n",
    "            intergene_seq = seq_record.seq[last_end:this_start]\n",
    "            if len(intergene_seq) > 1000: \n",
    "                intergene_seq = intergene_seq[-1000:]\n",
    "            intergenic_records.append(SeqRecord(intergene_seq,id=cds_list[ii][3],\n",
    "                         description='%d,%d,%d'%(last_end,this_start,strand)))                \n",
    "                    \n",
    "\n",
    "# STILL NEED TO TAKE CARE OF BOUNDARIES\n",
    "# REVERSE MODE NOT WORKING PROPERLY \n",
    "\n",
    "# reverse strand\n",
    "for ii, features in reversed(list(enumerate(cds_list))):               \n",
    "    strand = cds_list[ii][2]\n",
    "    if strand == -1: # only look downstream for intergenic region\n",
    "        if ii < len(cds_list)-1:\n",
    "            this_start = cds_list[ii][1]\n",
    "            this_start, last_end = get_rev_start_and_end(this_start,ii+1,strand)\n",
    "            \n",
    "            if last_end < this_start: \n",
    "                intergene_seq = seq_record.seq[this_start:genome_size] + seq_record.seq[0:last_end]\n",
    "            else:\n",
    "                intergene_seq = seq_record.seq[this_start:last_end]\n",
    "            if len(intergene_seq) > 1000:\n",
    "                intergene_seq = intergene_seq[0:1000]\n",
    "            intergenic_records.append(SeqRecord(intergene_seq.reverse_complement(),id=cds_list[ii][3],\n",
    "                         description='%d,%d,%d'%(last_end,this_start,strand)))                \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# place records in dataframe\n",
    "\n",
    "intergenic_record_dict = {'locus_tag':[],'start':[],'end':[],'strand':[],'seq':[]}\n",
    "for record in intergenic_records: \n",
    "    intergenic_record_dict['locus_tag'].append(record.id)\n",
    "    mystart, myend, strand = record.description.split(',')\n",
    "    mystart, myend, strand = int(mystart), int(myend), int(strand)\n",
    "    intergenic_record_dict['start'].append(mystart)\n",
    "    intergenic_record_dict['end'].append(myend)    \n",
    "    intergenic_record_dict['strand'].append(strand)    \n",
    "    intergenic_record_dict['seq'].append(str(record.seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "intergenic_record_df = pd.DataFrame(intergenic_record_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# intergenic_record_df.to_csv('intergenic_records.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
